### ISSUE TITLE
GPU memory requirements

### ISSUE THREAD

**Issue by xionluhnis** [HyeonwooNoh/DecoupledNet#2]
I am trying to fine-tune DecoupledNet with a specific dataset of mine but am failing to do so with Cuda.
The CPU version works, but it's not realistic given the time it takes, so I wonder if you had any
clear requirements about the GPU card that I may have missed.

I unsuccessfully tried the example of [002_train_seg_Full_anno](https://github.com/HyeonwooNoh/DecoupledNet/tree/master/training/002_train_seg_Full_anno), with a NVIDIA K20 (6GB).
- DecoupledNet_Full_anno is ~1.6GB
- fp_bp_model is ~1GB

[Here](https://github.com/HyeonwooNoh/DecoupledNet/files/171325/log.txt) is the training error.

For my dataset, I tried to split it (full ~20GB) into 1~2GB slices to train on, but I couldn't and got cuda out of memory. I even tried using a batch size of 1 with a set of only 100 pictures, and it still failed because of memory.

What GPU did you use for training? Shouldn't 6GB be sufficient?
Is caffe replicating the model in memory needlessly? What makes it use so much memory?

Thanks for any hint and/or hardware requirement.

_Note: I tried training on two different machines, each one with a Nvidia K20 (6GB), the result is the same._

**Comment by williford** [comment]
The authors (according to their paper) used the GTX Titan X GPU (12 GB).

The log file you give uses a batch size of 4, which on my system seems to use most of the 12 GB of memory. Do you set the batch size using the solver.prototxt? Because of the batch normalization, reducing the size of the batch might not be a good option.

**Comment by xionluhnis** [comment]
Mmh, I missed the 12G note. Thank you for that.
Currently, I have a K20 with only 6GB. Do you think it's possible to adapt either the dataset (what I tried without success) or the model to be trainable with only 6GB?
Is there a mode available to run on smaller configuration?

My full dataset is 20GB, so it won't fit in one training anyway with current hardware.
I am just surprised that using only a few hundred images as dataset brings me out of memory. I am probably missing something related to how Caffe works, but why would it need 3 copies of the full model during training?
I'm guessing 3 because right now the model takes around 2.4GB per instance (at least that's what it says during the network generation) and it crashes even though I have 6GB available, while I could get one iteration to go through with a very small dataset (then it crashed on the second iteration).

It seems to allocate memory for both the training and the testing network, even though they share all the same parameters (except for a different data source, and an extra precision layer - cutting that last one doesn't change anything since it has no parameters).

**Comment by williford** [comment]
You might be able to get a Titan X for free from NVIDIA for academic use, via: https://developer.nvidia.com/academic_hw_seeding
This is how I got my graphics card.

Have you calculated the size of the blobs and parameters needed for the model? You should be able to calculate the maximum batch size needed from this.

Did you make sure you are not running anything else on the graphics card when you run your model (`nvidia-smi`)?

**Comment by xionluhnis** [comment]
We indeed went with the Titan X and it works, but I'm surprised that when training with the same system on 002_train_seg_Full_anno, it gets to 10GB used on the GPU.
Seemingly, it may not be using all of it, since during the first few step, it only goes to 5GB, then on the next one to 8GB, and finally 10GB, and eventually saturate there.
I checked with `nvidia-smi` that I was not using GPU memory, and I am not.

Right now, with the Titan X, it's working, but my issue is that I'm planning to extend the network to look at multiple frames, and that will easily double the memory requirement, which then wouldn't fit on the Titan X either...
How easy would it be to port the current system to the newer versions of Caffe? I've heard that they optimized the memory usage by a lot since 2014 (when this fork was started, I think). Plus they have CuDNNv4 available now. So I'm tempted to make DecoupledNet work with the newer versions of Caffe.
